PTB and Fukui

When I attempt to run Fukui index calculations with PTB using the following command, I get an error 

**To Reproduce**
`xtb geom.xyz --ptb --vfukui`

`geom.xyz`
```
3

H 0 0 0
O 0 0 1
H 0 1 1
```

Error:
```
...
Fukui index Calculation

Run single point for reduced species

Run single point for oxidized species

Fukui functions:
     #        f(+)     f(-)     f(0)
     1H       0.000    0.000    0.000
     2O       0.000    0.000    0.000
     3H       0.000    0.000    0.000
########################################################################
[ERROR] Program stopped due to fatal error
-5- Calculation terminated
-4- ptb_calculator_singlepoint: PTB two-step SCF terminated
-3- ptb_calculator_singlepoint: Number of electrons and spin multiplicity do not match.
-2- ptb_calculator_singlepoint: PTB two-step SCF terminated
-1- ptb_calculator_singlepoint: Number of electrons and spin multiplicity do not match.
########################################################################
```

--- a/src/prog/main.F90
+++ b/src/prog/main.F90
@@ -550,7 +550,7 @@ contains
                p_run_modef, p_run_mdopt, p_run_metaopt)
             if (set%mode_extrun == p_ext_gfnff) then
                fnv = xfind(p_fname_param_gfnff)
-             elseif (set%mode_extrun .eq. p_ext_ptb) then
+            elseif (set%mode_extrun .eq. p_ext_ptb) then
                   fnv = xfind(p_fname_param_ptb)
             else
                if (set%gfn_method == 0) then
@@ -564,14 +564,18 @@ contains
                end if
             end if
          case (p_run_vip, p_run_vea, p_run_vipea, p_run_vfukui, p_run_vomega)
-            if (set%gfn_method == 0) then
-               fnv = xfind(p_fname_param_gfn0)
-            end if
-            if (set%gfn_method == 1) then
-               fnv = xfind(p_fname_param_gfn1)
-            end if
-            if (set%gfn_method == 2) then
-               fnv = xfind(p_fname_param_gfn2)
+            if (set%mode_extrun .eq. p_ext_ptb) then
+                  fnv = xfind(p_fname_param_ptb)
+            else
+               if (set%gfn_method == 0) then
+                  fnv = xfind(p_fname_param_gfn0)
+               end if
+               if (set%gfn_method == 1) then
+                  fnv = xfind(p_fname_param_gfn1)
+               end if
+               if (set%gfn_method == 2) then
+                  fnv = xfind(p_fname_param_gfn2)
+               end if
             end if
          end select
       end if

--- a/src/vertical.f90
+++ b/src/vertical.f90
@@ -30,14 +30,14 @@ module xtb_vertical
    private
 
    public :: vfukui
-   
+
 contains
 
 !--------------------------------------------------
 ! Calculate Fukui indices
 !--------------------------------------------------
 subroutine vfukui(env, mol, chk, calc, fukui)
-   
+
    implicit none
    !> Dummy-argument list
    class(TCalculator), intent(inout) :: calc
@@ -48,27 +48,52 @@ subroutine vfukui(env, mol, chk, calc, fukui)
       !! molecular information
    type(TRestart), intent(inout) :: chk
    type(TRestart) :: wf_an, wf_cat
-   
+
    ! fukui functions f(+), f(-), f(0)
    real(wp), intent(out) :: fukui(3,mol%n)
  
    type(scc_results) :: res
    real(wp) :: sigma(3,3)
-   real(wp) :: etot2,egap
+   real(wp) :: etot2,egap, orig_chrg
    real(wp) :: g(3,mol%n)
    logical :: exist
-   integer :: i
-   
+   integer :: i, orig_uhf
+
    write(env%unit,'(a)')
    write(env%unit,'("Fukui index Calculation")')
+
+   ! save original charge and uhf
+   orig_chrg = mol%chrg
+   orig_uhf = mol%uhf
 
    ! copy wavefunction
    wf_an%wfn = chk%wfn
    wf_cat%wfn = chk%wfn
 
    ! reduce the charge -> anion
-   mol%chrg = mol%chrg - 1       
-   if (mod(wf_an%wfn%nel,2).ne.0) wf_an%wfn%nopen = 1
+   mol%chrg = mol%chrg - 1
+   ! Create UHF entry for cation and anion -> we assume the UHF state to be equal for both species (adding or removing always towards the lowest multiplicity possible)
+   if (mod(wf_an%wfn%nel, 2) == 0 .and. (mol%uhf == 0)) then
+      ! if even number of electrons in the species (before charge modification), then we assume a closed-shell system
+      ! -> Increase UHF to 1
+      wf_an%wfn%nopen = 1
+      wf_cat%wfn%nopen = 1
+      ! mol object has to be modified as of today, since PTB sets its wavefunction only in the singlepoint calculation and checks for consistency with mol%uhf
+      mol%uhf = 1
+   elseif ((mod(wf_an%wfn%nel, 2) /= 0) .or. (mol%uhf > 0)) then
+      ! open-shell system (before charge modification)
+      ! -> Decrease UHF to 0
+      if ((wf_an%wfn%nopen == 0) .or. (mol%uhf == 0)) then
+         call env%terminate("Open shell system with closed shell wavefunction or UHF assignment.")
+         ! Wavefunction does not match number of electrons (and UHF entry)
+      end if
+      wf_an%wfn%nopen = wf_an%wfn%nopen - 1
+      wf_cat%wfn%nopen = wf_cat%wfn%nopen - 1
+      ! mol object has to be modified as of today, since PTB sets its wavefunction only in the singlepoint calculation and checks for consistency with mol%uhf
+      mol%uhf = mol%uhf - 1
+   else
+      call env%terminate("Invalid UHF and wavefunction combination.")
+   end if
 
    ! Perform single point calculation for anion
    write(env%unit,'(a)')
@@ -77,26 +102,25 @@ subroutine vfukui(env, mol, chk, calc, fukui)
 
    ! increase the charge -> cation
    mol%chrg = mol%chrg + 2
-   if (mod(wf_cat%wfn%nel,2).ne.0) wf_cat%wfn%nopen = 1
    ! Perform single point calculation for cation
    write(env%unit,'(a)')
    write(env%unit,'("Run single point for oxidized species")')
    call calc%singlepoint(env, mol, wf_cat, 1, exist, etot2, g, sigma, egap, res)
-  
+
    ! Calculate the fukui functions where N is the number of electrons
    ! see J. Am. Chem. Soc. 1986, 108, 19, 5708–5711 for equations
    ! keep in mind that their q are populations (p), 
    ! which are related to our q by q = Z-p 
 
-   ! f(+) = q_N -  q_(N+1) / neutral - anion
+   ! f(+) = q_(N_elec) -  q_(N_elec + 1) / neutral - anion
    fukui(1,:) = chk%wfn%q  - wf_an%wfn%q
-   
-   ! f(-) = q_(N-1) - q_N / cation - neutral
+
+   ! f(-) = q_(N_elec - 1) - q_(N_elec) / cation - neutral
    fukui(2,:) = wf_cat%wfn%q-chk%wfn%q
-   
-   ! f(0) = 0.5 * [q_(N-1) - q_(N+1)] / cation - anion
+
+   ! f(0) = 0.5 * [q_(N_elec - 1) - q_(N_elec + 1)] / cation - anion
    fukui(3,:) = 0.5d0*(wf_cat%wfn%q-wf_an%wfn%q)
-   
+
    ! Print out fukui functions
    write(env%unit,'(a)')
    write(env%unit,'("Fukui functions:")')
@@ -104,7 +128,9 @@ subroutine vfukui(env, mol, chk, calc, fukui)
    do i=1,mol%n
       write(env%unit,'(i6,a4,2f9.3,2f9.3,2f9.3)') i, mol%sym(i), fukui(1,i), fukui(2,i), fukui(3,i)
    enddo
-   mol%chrg = mol%chrg - 1
+   write(env%unit,'(a)')
+   mol%chrg = orig_chrg
+   mol%uhf = orig_uhf
 
 end subroutine vfukui
 

